### PR TITLE
tools: Enable `unicorn/prefer-type-error` rule and pre-apply fixes

### PR DESCRIPTION
**Pull request by Josmithr**
From the [rule documentation](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-type-error.md):

> This rule enforces you to throw a TypeError after a type checking if-statement, instead of a generic Error.
> 
> It's aware of the most commonly used type checking operators and identifiers like typeof, instanceof, .isString(), etc, borrowed from [ES2017](https://tc39.github.io/ecma262/), [Underscore](https://underscorejs.org/), [Lodash](https://lodash.com/), and [jQuery](https://jquery.com/). For a complete list of the recognized identifiers, please take a look at the [identifier-definition](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/prefer-type-error.js#L3).
> 
> The rule investigates every throw-statement which throws a generic Error. It will fail if the throw-statement is the only expression in the surrounding block and is preceeded by an if-statement whose condition consists of type-checks exclusively. You have to replace the Error with a TypeError.

Also update various usages of "new Error(foo)" to use `LoggingError` where possible.

### CODE REVIEW

**Comment by curtisman** [comment]
Just wondering how should I be thinking about using our `UsageError` vs `TypeError` if it is for guarding APIs.

Or more generally, I haven't pay attention to this, but it looks like we have an extensive "error" system.   How is `TypeError` fits in to it? 

@markfields FYI

**Comment by Josmithr** [comment]
> Just wondering how should I be thinking about using our `UsageError` vs `TypeError` if it is for guarding APIs.
> 
> Or more generally, I haven't pay attention to this, but it looks like we have an extensive "error" system. How is `TypeError` fits in to it?
> 
> @markfields Mark Fields FTE FYI

Note that this exclusively checks for *generic* `Error`s being thrown, so it will not have any opinion if we throw custom error types in these cases.

**Review by tylerbutler — APPROVED** [review]
👍🏻

**Comment by Josmithr** [comment]
There appears to be an issue with full symlink builds currently. It appears I have not actually validated the code against this new rule yet. I'm going to hold off on merging this until I've had a chance to validate.

File: experimental/PropertyDDS/packages/property-common/src/datastructures/integer64.ts [hunk 1]
```diff
@@ -100,6 +100,7 @@ function _stringToInt64(in_signed: boolean, in_string: string, in_radix = 10): n
     while (position < string.length) {
         const digit = parseInt(string[position++], in_radix);
         if (isNaN(digit)) {
+            // eslint-disable-next-line unicorn/prefer-type-error
```
Comment (tylerbutler) [inline experimental/PropertyDDS/packages/property-common/src/datastructures/integer64.ts:103]: What's special about this one?

**Review by tylerbutler — APPROVED** [review]

File: experimental/PropertyDDS/packages/property-common/src/datastructures/integer64.ts (hunk 1 above)
Comment (tylerbutler) [reply experimental/PropertyDDS/packages/property-common/src/datastructures/integer64.ts:103]: Same question for the other couple of exceptions.

Comment (Josmithr) [reply experimental/PropertyDDS/packages/property-common/src/datastructures/integer64.ts:103]: I didn't look closely enough. I missed that this was checking the result of string parsing. This isn't special; I'll update.

**Comment by markfields** [comment]
This is great to see `throw new Error` getting cleaned up!  Preferable to `TypeError` - at least based on the current setup - would be to use `LoggingError`, or a class that extends it (e.g. `UsageError` may apply in many of these cases as Curtis mentioned - although whether "usage error" applies broadly throughout the codebase is debatable).

If you want more context on how stuff works, I just wrote up an issue AB#693 which tries to explain the current approach and points out area for improvement.  In light of this PR, I might add to that document - aligning more with industry standards like `TypeError` could also be a virtuous direction to consider, but it just needs to be harmonious with existing system (assumptions/values).  Today we aren't leverage `Error.name` at all, but maybe we "should".

Anyway, back to this PR - my request would be to go straight to `LoggingError` as the default here, and not just for `throw new Error` under an `if` statement but anywhere in the code (well, anywhere with a dependency on `telemetry-utils` package).

**Review by markfields — APPROVED** [review]
🎂 LGTM!

**Pull request merged**